### DISCUSSION TITLE
How to enable merged model responses?

### DISCUSSION THREAD

**Discussion by andrewginns** [open-webui/open-webui#4934] [Q&A] [answered]
I see that the feature is included in the [changelog for 0.3.14](https://github.com/open-webui/open-webui/blob/main/CHANGELOG.md#0314---2024-08-21).
> 🔀 Merge Responses in Many Model Chat: Enhances the dialogue by merging responses from multiple models into a single, coherent reply, improving the interaction quality in many model chats

I can't find the controls for this in the UI. When I enable 2 models in chat I just get the expected 2 outputs. Is there something I'm missing?
<img width="740" alt="image" src="https://github.com/user-attachments/assets/9241021d-979c-477e-a1a5-48b1eeeb5827">

**Comment by andrewginns** [comment, marked as answer]
I realise it was hiding in plain sight, this button merges the responses of the multiple models.
![image](https://github.com/user-attachments/assets/9c417257-95b0-4ddd-9e8a-8d826fac78fa)

**Comment by djdeniro** [comment]
anyway, who provide merged response? how it works?

**Reply by marschr** [reply]
I was asking myself the same, looking from the ollama server logs (with `OLLAMA_DEBUG=1`) and after a few runs with multiple models it seems that the last model on the list of selected models for the current chat is being used to combine, in my case `smollm2:135m`:
![Screenshot from 2024-11-01 15-38-36](https://github.com/user-attachments/assets/7e205c72-2355-428d-a428-8d917ebf4a1f)
 
 I cound't find much info about it on the documentation though.
 
Also, here's the initial part of the prompt used to generate the merges:
> prompt="<|im_start|>system\nYou are a helpful AI assistant named SmolLM, trained by Hugging Face<|im_end|>\n<|im_start|>user\nYou have been provided with a set of responses from various models to the latest user query: \"Why is the sky blue?\"\n\nYour task is to synthesize these responses into a single, high-quality response. It is crucial to critically evaluate the information provided in these responses, recognizing that some of it may be biased or incorrect. Your response should not simply replicate the given answers but should offer a refined, accurate, and comprehensive reply to the instruction. Ensure your response is well-structured, coherent, and adheres to the highest standards of accuracy and reliability.\n\nResponses from models: \"\"\"The sky appears blue because of a phenomenon called Rayleigh scattering, [...]

The prompt is defined at https://github.com/open-webui/open-webui/blob/7228b39064ac28e1240bf8998f2a35535c6f7ef5/backend/open_webui/main.py#L1836 in the `generate_moa_response` method (moa = mixture of agents). The code references a `task_model_id` as being the one used for the merging prompt call, but I didn't find a direct definition of what a `task_model_id` is in the open-webui context.
Hope someone finds this insightful too.

**Reply by miloslavnosek** [reply]
> I was asking myself the same, looking from the ollama server logs (with `OLLAMA_DEBUG=1`) and after a few runs with multiple models it seems that the last model on the list of selected models for the current chat is being used to combine, in my case `smollm2:135m`: ![Screenshot from 2024-11-01 15-38-36](https://private-user-images.githubusercontent.com/5367013/382374196-7e205c72-2355-428d-a428-8d917ebf4a1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwNDkxNDEsIm5iZiI6MTczMjA0ODg0MSwicGF0aCI6Ii81MzY3MDEzLzM4MjM3NDE5Ni03ZTIwNWM3Mi0yMzU1LTQyOGQtYTQyOC04ZDkxN2ViZjRhMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMjA0MDQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2VlNjQwOTFlMTAwNTJjMDFhMDQxNTY5ZTBjMDMwY2FlZGE0NmI1NWE3NThkY2JhYjJkNjc5NGY5NzFhODk0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2PSB_YFEXzV1w2Y_n8FZl_b8B3M4IBVJayE3iMXzrBg)
> 
> I cound't find much info about it on the documentation though.
> 
> Also, here's the initial part of the prompt used to generate the merges:
> 
> > prompt="<|im_start|>system\nYou are a helpful AI assistant named SmolLM, trained by Hugging Face<|im_end|>\n<|im_start|>user\nYou have been provided with a set of responses from various models to the latest user query: "Why is the sky blue?"\n\nYour task is to synthesize these responses into a single, high-quality response. It is crucial to critically evaluate the information provided in these responses, recognizing that some of it may be biased or incorrect. Your response should not simply replicate the given answers but should offer a refined, accurate, and comprehensive reply to the instruction. Ensure your response is well-structured, coherent, and adheres to the highest standards of accuracy and reliability.\n\nResponses from models: """The sky appears blue because of a phenomenon called Rayleigh scattering, [...]
> 
> The prompt is defined at
> 
> https://github.com/open-webui/open-webui/blob/7228b39064ac28e1240bf8998f2a35535c6f7ef5/backend/open_webui/main.py#L1836
> in the `generate_moa_response` method (moa = mixture of agents). The code references a `task_model_id` as being the one used for the merging prompt call, but I didn't find a direct definition of what a `task_model_id` is in the open-webui context.
> Hope someone finds this insightful too.

Thank you. I wish this information were clarified somewhere. I wanted to know which model is used for merging to make a decision about which one could potentially do the best job, but I couldn't find it in either the documentation or the icon tooltip.
